designdocs/batched_messages.md

```diff
@@ -0,0 +1,81 @@
+# Batched Messages
+
+## Table of Contents
+* [Context and Scope](#context-and-scope)
+* [Problem Statement](#problem-statement)
+* [Goals and Non-Goals](#goals-and-non-goals)
+* [Proposal](#proposal)
+* [Benefits and Risks](#benefits-and-risks)
+* [Alternatives and Rationale](#alternatives-and-rationale)
+* [Glossary](#glossary)
+
+## Context and Scope
+This design document describes batched messages and APIs for sending and retrieving them.
+
+## Problem Statement
+In order to reduce costs for announcers of messages on-chain as well as reduce network congestion, announcers collate messages into batches of the same type of message and announce the batch location on-chain, instead of announcing each individual message.
+
+## Goals and Non-Goals
+This specifies how batches are to be announced on chain: what is required and how a batch may be partially verified based on on-chain information.
+
+This document does not describe the messages that are supported for batch announcements. Please see the [DSNP Specification](https://spec.dsnp.org/).
+
+This document does not discuss validation of batch file format located at the URL in the announcement, since
+the file format cannot be verified on-chain. For details about batch file format, see [DSNP Spec: Batch Publications](https://spec.dsnp.org/DSNP/BatchPublications).
+
+## Proposal
+* All names are placeholders and may be changed.
+* Errors in the extrinsic(s) must have different, reasonably-named error enums for each type of error for ease of debugging.
+
+### Constants
+* `BatchPageSizeMax` - the maximum number of batch announcements that will be returned by the `get_batches` query
+* `BatchSizeMinBytes` - the minimum possible size of a valid batch file with 1 row
+
+### Enums
+* `MessageType` - all values specified in [DSNP Spec Announcement Types](https://spec.dsnp.org/DSNP/Announcements.html)
+
+### Types
+* `BatchAnnouncementParams<T:Config>`: generic
+    * `batch_url`:`Vec<u8>` the URL of the batch file.  Must be IPFS or HTTPS URL. The URL must be correctly formatted.
+    * `message_type`: `MessageType`  the type of messages in this batch
+    * `file_size`: `usize`, the size of the batch file, used to determine message fee.  Must be &gt;= the minimum possible DSNP batch file size.
+    * `file_hash`: `<T::Hash>`, the hash of the batch file. Must not be 0 hash
+    * `msa_id`: `MsaId`, the id to use for the announcer.  Must exist and be active
+
+* `BatchAnnouncement`: implements `BatchAnnouncementParams`, returned by `get_batches`
+
+* `BatchAnnouncementOptions<T:Config>`: struct
+    * `msa_id`:  `MsaId`, the announcer's MSA id.  If 0, get all announcements
+    * `block_number`: `<T::BlockNumber>`, retrieve messages starting at the given block number (inclusive). If 0, get all messages
```

Comment:
I think using default of any param if not set (here `0`) to get all is dangerous. I don't think there should be a get all option since depend on the size stored it can be huge.

For pagination via block number I've defined a request and response. Please take a look if makes sense to use something like that.
https://github.com/LibertyDSNP/mrc/blob/main/common/primitives/src/messages.rs#L26-L58

Reply:
I have some questions about your implementation and also about paging data from the db but it'll be quicker to talk about it in Zoom.

Reply:
Sure. Let me know